src/handler.ts

```diff
@@ -0,0 +1,299 @@
+
+
+/**
+ * A type that defines a handler for a given operation.
+ */
+export type OperationHandlerFor<T> =
+  T extends Operation<infer I, infer O> ? OperationHandler<I, O> | SyncOperationHandler<I, O> : never;
+
+/**
+ * A type that defines a collection of handlers for a given collection of operation interfaces.
+ */
+export type ServiceHandlerFor<T extends OperationMap = OperationMap> = {
+  [K in keyof T & string]: OperationHandlerFor<T[K]>;
+};
+
+/**
+ * A {@link Service} that includes a collection of handlers for its operations.
+ */
+export interface ServiceHandler<T extends OperationMap = OperationMap> extends Service<T> {
+  handlers: ServiceHandlerFor<T>;
+}
+
+/**
+ * Constructs a service handler for a given service contract.
+ */
+export function serviceHandler<T extends OperationMap>(
+  service: Service<T>,
+  handlers: ServiceHandlerFor<T>,
+): ServiceHandler<T> {
+  const ops = new Set<string>();
+
+  for (const [k, op] of Object.entries(service.operations)) {
+    if (!op.name) {
+      throw new TypeError(`Tried to register an operation with no name for service ${service.name} with key ${k}`);
+    }
+    if (ops.has(op.name)) {
+      throw new TypeError(`Operation with name ${op.name} already registered for service ${service.name}`);
+    }
+    const handler = handlers[k];
+    if (!handler) {
+      throw new TypeError(`No handler registered for ${k} on service ${service.name}`);
+    }
+    ops.add(op.name);
+  }
+
+  return {
+    ...service,
+    handlers,
+  };
+}
+
+/**
+ * A collection of service handlers that dispatches requests to the registered service and operation handler.
+ */
+export class ServiceRegistry implements OperationHandler<unknown, unknown> {
+  private services = new Map<string, Map<string, OperationHandler<any, any> | SyncOperationHandler<any, any>>>();
+
+  constructor(services: ServiceHandler[]) {
+    for (const s of services) {
+      if (!s.name) {
+        throw new TypeError("Tried to register a Nexus service with no name");
+      }
+      if (this.services.has(s.name)) {
+        throw new TypeError(`Duplicate registration of nexus service ${s.name}`);
+      }
+      const ops = new Map<string, OperationHandler<any, any> | SyncOperationHandler<any, any>>();
+      for (const [k, op] of Object.entries(s.operations)) {
+        if (!op.name) {
+          throw new TypeError(`Tried to register an operation with no name for service ${s.name} with key ${k}`);
+        }
+        if (ops.has(op.name)) {
+          throw new TypeError(`Operation with name ${op.name} already registered for service ${s.name}`);
+        }
+        const handler = s.handlers[k];
+        if (!handler) {
+          throw new TypeError(`No handler registered for ${k} on service ${s.name}`);
+        }
+        ops.set(op.name, handler);
+      }
+      this.services.set(s.name, ops);
+    }
+  }
+
+  private getHandler(ctx: OperationContext): OperationHandler<any, any> | SyncOperationHandler<any, any> {
```

Comment:
Term "Handler" is a bit overloaded. Let's be more specific.

```suggestion
  private getOperationHandler(ctx: OperationContext): OperationHandler<any, any> | SyncOperationHandler<any, any> {
```